core/src/main/java/org/elasticsearch/bootstrap/BootstrapCLIParser.java

```diff
@@ -0,0 +1,168 @@
+
+        private static final CliToolConfig.Cmd CMD = cmd(NAME, Start.class)
+                .options(
+                        optionBuilder("d", "daemonize").hasArg(false).required(false),
+                        optionBuilder("p", "pidfile").hasArg(true).required(false),
+                        optionBuilder("V", "version").hasArg(false).required(false),
+                        Option.builder("D").argName("property=value").desc("use value for given property").valueSeparator('=').numberOfArgs(2).argName("D")
+                )
+                .lenient(true) // needed to parse the voodoo opts I guess
+                .build();
+
+        public static Command parse(Terminal terminal, CommandLine cli) {
+            if (cli.hasOption("V")) {
+                return Version.parse(terminal, cli);
+            }
+
+            System.setProperty("es.foreground", "true");
+            if (cli.hasOption("d")) {
+                System.clearProperty("es.foreground");
+            }
+
+            String pidFile = cli.getOptionValue("pidfile");
+            if (!Strings.isNullOrEmpty(pidFile)) {
+                System.setProperty("es.pidfile", pidFile);
+            }
+
+            if (cli.hasOption("D")) {
+                Properties properties = cli.getOptionProperties("D");
+                for (Map.Entry<Object, Object> entry : properties.entrySet()) {
+                    System.setProperty("es." + entry.getKey().toString(), entry.getValue().toString());
+                }
+            }
+
+            // hacky way to extract all the fancy extra args, there is no CLI tool helper for this
+            Iterator<String> iterator = cli.getArgList().iterator();
+            while (iterator.hasNext()) {
+                String arg = iterator.next();
+                if (!arg.startsWith("--")) {
+                    throw new IllegalArgumentException("Parameter [" + arg + "]does not start with --");
+                }
+                // if there is no = sign, we have to get the next argu
+                arg = arg.replace("--", "");
+                if (arg.contains("=")) {
+                    String[] splitArg = arg.split("=", 2);
+                    String key = splitArg[0];
+                    String value = splitArg[1];
+                    System.setProperty("es." + key, value);
+                } else {
+                    if (iterator.hasNext()) {
+                        iterator.remove();
+                        String value = iterator.next();
+                        if (value.startsWith("--")) {
+                            throw new IllegalArgumentException("Parameter [" + arg + "] needs value");
+                        }
+                        System.setProperty("es." + arg, value);
+                    } else {
+                        throw new IllegalArgumentException("Parameter [" + arg + "] needs value");
+                    }
+                }
+                iterator.remove();
```

Comment:
very minor, but o we need to remove form the iterator here? seems like it wouldn't matter if we left the entry in the list?